### DISCUSSION TITLE
Need help to understand and use this

### DISCUSSION THREAD

**Discussion by mrdevver** [guest271314/persistent-serviceworker#2] [General]
Hi guest271314,

After researching a lot on internet, I ended up in this repo, look like you resolved the problem of impersistent service worker in chrome extension's background.

Could you please help to guide me or explain how to implement your solution into our extension?

Thanks a lot

**Comment by guest271314** [comment]
Sure. What issues are you having implementing the code in this repository?

**Reply by mrdevver** [reply]
Hi guy,

Sorry for the dumb question, my extension is experiencing this issue that the background died after 5 mins, look like you resolved this but I don't know how to apply your solution to my project, could you please explain a little bit, tell me what I need to do to bring it to my project? Which file, which code should I need to check ?

Thank you so much.

**Reply by guest271314** [reply]
You can use this version https://github.com/guest271314/persistent-serviceworker/tree/main/chromium_extension_web_accessible_resources_iframe_message_event.

Append a `<iframe>` to the target `document` with `src` set to an HTML document defined in the extension. From the `<iframe>` which becomes a `WindowClient` post messages to the `ServiceWorker` every 15 seconds with `(await navigator.serviceWorker.ready).active.postMessage(null);`.

**Reply by guest271314** [reply]
It looks like Chromium 112 is no longer registering `iframe` with `src` set to a `chrome-extension:` protocol as a `WindowClient`. Appears counterintuitive to me. There are ways to work around that nonetheless. Using `"externally_connectable"` approach still works. You will need to define the sites to match which can connect to the `ServiceWorker` in `"matches"` array in manifest.json https://github.com/guest271314/persistent-serviceworker/tree/main/chromium_extension_externally_connectable.

**Reply by guest271314** [reply]
The cause of https://github.com/guest271314/persistent-serviceworker/tree/main/chromium_extension_web_accessible_resources_iframe_message_event not working as expected on Chromium 112 is evidently rooted in at chrome://flags

> Experimental third-party storage partitioning.
> Enables partitioning of third-party storage by top-level site. Note: this is under active development and may result in unexpected behavior. Please file bugs at https://bugs.chromium.org/p/chromium/issues/entry?labels=StoragePartitioning-trial-bugs&components=Blink%3EStorage. – Mac, Windows, Linux, ChromeOS, Android, Fuchsia, Lacros
> 
> [#third-party-storage-partitioning](chrome://flags/#third-party-storage-partitioning)

Either disable at chrome://flags or launch with `--disable-features=ThirdPartyStoragePartitioning` flag and the message event approach works as expected.

**Reply by mrdevver** [reply]
Hi @guest271314 ,

In my extension, i need to keep background service worker alive to handle message from my content script, so I think we cannot ask users to  change settings in chrome://flags, they won't know how to do.

> Append a <iframe> to the target document with src set to an HTML document defined in the extension. From the <iframe> which becomes a WindowClient post messages to the ServiceWorker every 15 seconds with (await navigator.serviceWorker.ready).active.postMessage(null);.

I'm not sure how to implement this as your guideline but let me try, thank you.

**Reply by guest271314** [reply]
Basically you just need to keep some form of communication going with the `ServiceWorker` or use a Native Messaging host.

A never ending `fetch` request and `EventSource` from a `WindowClient` keeps `ServiceWorker` active indefinitely. So too will using `"externally_connectable"` and passing messages back and forth ever 15 seconds or so.

**Reply by mrdevver** [reply]
Hi @guest271314 ,

The externally_connectable you mean is this piece https://github.com/guest271314/persistent-serviceworker/tree/main/chromium_extension_externally_connectable , is it correct ?

I just need to bring the code in background.js to my background.js, right ? And because I have a popup open when clicking the extension icon, so instead of executing the func persistentExternalConnection in chrome.action.onClicked , can I execute it when load the content scipt ?

Thank you.
